_data/meltano/extractors/tap-nice-incontact/singer-io.yml

```diff
@@ -0,0 +1,87 @@
+capabilities:
+- catalog
+- discover
+- state
+description: NICE Systems provides intent-based solutions that capture and analyze
+  client interactions and transactions for businesses.
+domain_url: https://www.nice.com/
+keywords:
+- api
+label: NICE inContact
+logo_url: /assets/logos/extractors/nice-incontact.png
+maintenance_status: active
+name: tap-nice-incontact
+namespace: tap_nice_incontact
+next_steps: ''
+pip_url: git+https://github.com/singer-io/tap-nice-incontact.git
+repo: https://github.com/singer-io/tap-nice-incontact
+settings:
+- description: NICE inContact API version. Default is "21.0"
+  label: Api Version
+  name: api_version
+- description: 'Process and email for API logging purposes. Example: tap-nice-incontact
+    <api_user_email@your_company.com>'
+  label: User Agent
+  name: user_agent
+- description: NICE inContact API key (see Authentication in usage section for more
+    details)
+  kind: password
+  label: API Key
+  name: api_key
+- description: 'NICE inContact instance cluster. Example: "c42"'
+  label: API Cluster
+  name: api_cluster
+- description: The NICE inContact auth domain/region to use. Default is "na1". See
+    Authentication for more.
+  label: Auth Domain
+  name: auth_domain
+- description: NICE inContact API secret (see Authentication in usage section for
+    more details)
+  kind: password
+  label: API Secret
+  name: api_secret
+- description: RFC3339 date string "2017-01-01T00:00:00Z"
+  kind: date_iso8601
+  label: Start Date
+  name: start_date
+- description: |
+    Stream specific reporting periods. The config is a mapping of stream name to reporting period, the options are: days, hours, and minutes.
+    Default days. The days/hours are in increments of 1 and minutes are in increments of 5 minutes.
+
+    ```yaml
+      periods
+        - skills_summary: days
+        - skills_sla_summary: days
+        - teams_performance_total: days
+        - wfm_skills_agent_performance: days
+        - wfm_agents: days
+        - wfm_agents_scorecards: days
+    ```
+
+    or using dot notation:
+
+    ```yaml
+      periods.skills_summary: days
+      periods.skills_sla_summary: days
+      periods.teams_performance_total: days
+      periods.wfm_skills_agent_performance: days
+      periods.wfm_agents: days
+      periods.wfm_agents_scorecards: days
+
+    ```
+  kind: object
+  label: Periods
+  name: periods
```

Comment:
@pnadolny13 assuming the streams are reasonably stable, would it make sense to add individual settings for each one?

```yaml
- name: periods.contacts_completed
  kind: options
  options:
  - label: Days
    value: days
  - label: Hours
    value: hours
  - label: Minutes
    value: minutes
- name: periods.skills_summary
  kind: options
  options:
  - label: Days
    value: days
  - label: Hours
    value: hours
  - label: Minutes
    value: minutes
- name: periods.skills_sla_summary
  kind: options
  options:
  - label: Days
    value: days
  - label: Hours
    value: hours
  - label: Minutes
    value: minutes
- name: periods.teams_performance_total
  kind: options
  options:
  - label: Days
    value: days
  - label: Hours
    value: hours
  - label: Minutes
    value: minutes
- name: periods.wfm_skills_contacts
  kind: options
  options:
  - label: Days
    value: days
  - label: Hours
    value: hours
  - label: Minutes
    value: minutes
- name: periods.wfm_skills_dialer_contacts
  kind: options
  options:
  - label: Days
    value: days
  - label: Hours
    value: hours
  - label: Minutes
    value: minutes
- name: periods.wfm_skills_agent_performance
  kind: options
  options:
  - label: Days
    value: days
  - label: Hours
    value: hours
  - label: Minutes
    value: minutes
- name: periods.wfm_agents
  kind: options
  options:
  - label: Days
    value: days
  - label: Hours
    value: hours
  - label: Minutes
    value: minutes
- name: periods.wfm_agents_schedule_adherence
  kind: options
  options:
  - label: Days
    value: days
  - label: Hours
    value: hours
  - label: Minutes
    value: minutes
- name: periods.wfm_agents_scorecards
  kind: options
  options:
  - label: Days
    value: days
  - label: Hours
    value: hours
  - label: Minutes
    value: minutes
```

It's very verbose but it allows users to set each of them with environment variables and they still can set them in a nested object in YAML.

Wdyt?

Reply:
> are they not able to set dot notation settings using environment variables if they arent a defined setting?

@pnadolny13 `object` settings can only be set with env vars with a full JSON string, e.g.

```shell
TAP_NICE_INCONTACT_PERIODS='{"skills_summary": "days", "skills_sla_summary": "minutes", ...}'
```

But if you know the object keys and you break them down as above then you can set each individually:

```shell
TAP_NICE_INCONTACT_PERIODS_SKILLS_SUMMARY='days'
TAP_NICE_INCONTACT_PERIODS_SKILLS_SLA_SUMMARY='minutes'
```

> So have them duplicated makes that a little ugly, maybe we can put that once in the usage section section though.

Totally agree, it would be a bit ugly to have that for each nested setting in the plugin page. Bit of a balance between Meltano UX and Hub presentation.

Reply:
@edgarrmondragon thanks for the context, I didnt know thats how it worked. I think this is a good thing to discuss further. I'm going to merge this PR as is for a first iteration then open a discussion to talk about how this topic more.